src/items/recipes.zig

```diff
@@ -0,0 +1,226 @@
+const items = main.items;
+const ZonElement = main.ZonElement;
+const NeverFailingAllocator = main.heap.NeverFailingAllocator;
+const NeverFailingArenaAllocator = main.heap.NeverFailingArenaAllocator;
+const Tag = main.Tag;
+const Recipe = items.Recipe;
+const BaseItemIndex = items.BaseItemIndex;
+const Block = main.blocks.Block;
+
+const Segment = union(enum) {literal: []const u8, symbol: []const u8};
+
+fn parsePattern(allocator: NeverFailingAllocator, pattern: []const u8) ![]Segment {
+	var segments: main.List(Segment) = .init(allocator);
+	defer segments.deinit();
+	var idx: usize = 0;
+	while(idx < pattern.len) {
+		if(pattern[idx] == '{') {
+			if(segments.items.len > 0 and segments.items[segments.items.len - 1] == .symbol) {
+				return error.AmbiguousSymbols;
+			}
+			idx += 1;
+			const endIndex = std.mem.indexOfScalarPos(u8, pattern, idx, '}') orelse return error.UnclosedBraces;
+			if(idx == endIndex) return error.EmptyBrackets;
+			const symbol = pattern[idx..endIndex];
+			segments.append(.{.symbol = symbol});
+			idx = endIndex + 1;
+		} else {
+			const endIndex = std.mem.indexOfScalarPos(u8, pattern, idx, '{') orelse pattern.len;
+			segments.append(.{.literal = pattern[idx..endIndex]});
+			idx = endIndex;
+		}
+	}
+	return segments.toOwnedSlice();
+}
+
+const ItemStackPattern = struct {
+	amount: u16,
+	pattern: []Segment,
+};
+
+fn parseItemZon(allocator: NeverFailingAllocator, zon: ZonElement) !ItemStackPattern {
+	var id = zon.as([]const u8, "");
+	var amount: u16 = 1;
+	if(std.mem.indexOfScalar(u8, id, ' ')) |index| blk: {
+		amount = std.fmt.parseInt(u16, id[0..index], 0) catch break :blk;
+		id = id[index + 1 ..];
+	}
+	const pattern = try parsePattern(allocator, id);
+	return .{
+		.amount = amount,
+		.pattern = pattern,
+	};
+}
+
+fn matchWithKeys(allocator: NeverFailingAllocator, target: []const u8, pattern: []const Segment, keys: *const std.StringHashMap([]const u8)) ![]std.StringHashMap([]const u8) {
+	var idx: usize = 0;
+	var newKeys = keys.clone() catch unreachable;
+	errdefer newKeys.deinit();
+
+	for(0.., pattern) |i, segment| {
+		switch(segment) {
+			.literal => |literal| {
+				if(literal.len + idx > target.len or !std.mem.eql(u8, target[idx .. idx + literal.len], literal)) {
+					return error.NoMatch;
+				}
+				idx += literal.len;
+			},
+			.symbol => |symbol| {
+				var indexes: main.List(usize) = .init(allocator);
+				defer indexes.deinit();
+				if(i + 1 < pattern.len) {
+					const nextSegment = pattern[i + 1];
+					var nextIndex = idx;
+					while(std.mem.indexOfPos(u8, target, nextIndex, nextSegment.literal)) |endIndex| {
+						indexes.append(endIndex);
+						nextIndex = endIndex + 1;
+					}
+				} else {
+					indexes.append(target.len);
+				}
+				if(indexes.items.len == 0) {
+					return error.NoMatch;
+				}
+				if(indexes.items.len == 1) {
+					if(newKeys.get(symbol)) |value| {
+						if(!std.mem.eql(u8, target[idx..indexes.items[0]], value)) {
+							return error.NoMatch;
```

Comment:
This check should be done in both branches, and thus I'd suggest putting it into the while loop above, before even adding the candidate to the indices list.